File: codeforces/edu/ex269103A/main.py
def solve(s, debug=0):
    n = len(s)
    nn = n + 1
    _ = lambda ls, i, k: ls[i] if i < k else 0  # Safe getter with 0 for out-of-bound

    # Convert to numeric
    s = [ord(c) + 1 for c in s]

    # Suffix index which we will sort
    a = list(range(nn))

    # Prefix score
    b = [0] * nn

    # Initialize (1-prefix)
    a.sort(key=lambda i: _(s, i, n))
    e = _(s, a[0], n)
    score = 0
    for i in range(1, nn):
        e_i = _(s, a[i], n)
        if e_i != e:
            e = e_i
            score += 1
        b[a[i]] = score

    if debug:
        print("m: 1,", a, b)

    # Recursion (m-prefix to 2m-prefix)
    m = 1
    a_tmp = [0] * nn
    b_tmp = [0] * nn
    counts = [0] * nn
    counts_cumsum = [0] * (nn + 1)
    while m < n:
        #
        # 1. Sort 2m-prefix using m-prefix score
        #

        # 1.1. Counting sort for [m, 2m]-fix

        # 1.1.1. counting
        for i in range(nn):
            score = _(b, a[i] + m, nn)  # @@@ a[i] + m @@@
            counts[score] += 1

        # 1.1.2. cumsum
        counts_cumsum[0] = 0
        for i in range(nn):
            counts_cumsum[i + 1] = counts[i] + counts_cumsum[i]
            counts[i] = 0  # reset for next use

        # 1.1.3. replacing
        for i in range(nn):
            score = _(b, a[i] + m, nn)  # @@@ a[i] + m @@@
            a_tmp[counts_cumsum[score]] = a[i]  # save to temporary
            counts_cumsum[score] += 1

        # 1.2. Counting sort for [0, m]-fix

        # 1.2.1. counting
        for i in range(nn):
            score = _(b, a_tmp[i], nn)  # @@@ a_tmp[i] @@@
            counts[score] += 1

        # 1.2.2. cumsum
        counts_cumsum[0] = 0
        for i in range(nn):
            counts_cumsum[i + 1] = counts[i] + counts_cumsum[i]
            counts[i] = 0

        # 1.2.3. replacing
        for i in range(nn):
            score = _(b, a_tmp[i], nn)  # @@@ a_tmp[i] @@@
            a[counts_cumsum[score]] = a_tmp[i]
            counts_cumsum[score] += 1

        # 2. Make 2m-prefix score using m-prefix score
        e = (_(b, 0, nn), _(b, m, nn))
        score = 0
        for i in range(1, n + 1):
            e_i = (_(b, a[i], nn), _(b, a[i] + m, nn))
            if debug >= 2:
                print(f"i: {i},", a[i], e_i)
            if e_i != e:
                e = e_i
                score += 1
            b_tmp[a[i]] = score
        b, b_tmp = b_tmp, b

        m = 2 * m
        if debug:
            print(f"m: {m},", a, b)

    return a


def main(istr, ostr):
    s = istr.readline().strip()
    print(*solve(s), file=ostr)


if __name__ == "__main__":
    import sys

    main(sys.stdin, sys.stdout)


import unittest
import io, sys


class Test(unittest.TestCase):
    def test_0(self):
        s = "ababba"
        print(solve(s, debug=1))

    def test_1(self):
        inp = """\
ababba
"""
        main(io.StringIO(inp), sys.stdout)

    def test_2(self):
        inp = """\
ppppplppp
"""
        main(io.StringIO(inp), sys.stdout)

File: codeforces/edu/ex269103A/main.js
//
// Rewrite main.py in javascript
//

const solve = (s, debug = 0) => {
  const n = s.length;
  const nn = n + 1;

  // Safe getter with 0 for out-of-bound
  const _ = (ls, i, k) => i < k ? ls[i] : 0;

  // Convert to numeric
  s = new TextEncoder().encode(s)

  // Suffix indices which we will sort
  const a = new Int32Array(nn);
  for (let i = 0; i < nn; i++) {
    a[i] = i;
  }

  // Prefix score
  let b = new Int32Array(nn)

  // Initialization (1-prefix)
  a.sort((i, j) => _(s, i, n) - _(s, j, n));
  {
    let e = _(s, a[0], n);
    let score = 0;
    for (let i = 0; i < nn; i++) {
      let e_i = _(s, a[i], n);
      if (e_i !== e) {
        e = e_i;
        score++;
      }
      b[a[i]] = score;
    }
  }

  if (debug) {
    console.log('m: 1, a:', ...a, 'b:', ...b);
  }

  // Recursion (m-prefix to 2m-prefix)
  let m = 1;
  let a_tmp = new Int32Array(nn);
  let b_tmp = new Int32Array(nn);
  let counts = new Int32Array(nn);
  let counts_cumsum = new Int32Array(nn + 1);
  while (m < n) {
    // 1. Sort 2m-prefix using m-prefix score by counting sort twice
    // 1.1. Counting sort (m, 2m)-fix
    for (let i = 0; i < nn; i++) {
      let score = _(b, a[i] + m, nn); // a[i] + m
      counts[score]++;
    }
    counts_cumsum[0] = 0;
    for (let i = 0; i < nn; i++) {
      counts_cumsum[i + 1] = counts[i] + counts_cumsum[i];
      counts[i] = 0;
    }
    for (let i = 0; i < nn; i++) {
      let score = _(b, a[i] + m, nn);
      a_tmp[counts_cumsum[score]] = a[i]; // a_tmp <-- a
      counts_cumsum[score] += 1;
    }
    // 1.1. Counting sort (0, m)-fix
    for (let i = 0; i < nn; i++) {
      let score = _(b, a_tmp[i], nn); // a_tmp[i]
      counts[score]++;
    }
    counts_cumsum[0] = 0;
    for (let i = 0; i < nn; i++) {
      counts_cumsum[i + 1] = counts[i] + counts_cumsum[i];
      counts[i] = 0;
    }
    for (let i = 0; i < nn; i++) {
      let score = _(b, a_tmp[i], nn);
      a[counts_cumsum[score]] = a_tmp[i]; // a <-- a_tmp
      counts_cumsum[score] += 1;
    }

    // 2. Make 2m-prefix score using m-prefix score
    {
      let e1 = _(b, a[0], nn);
      let e2 = _(b, a[0] + m, nn);
      let score = 0;
      for (let i = 1; i < nn; i++) {
        let e1_i = _(b, a[i], nn);
        let e2_i = _(b, a[i] + m, nn);
        if (e1_i !== e1 || e2_i !== e2) {
          e1 = e1_i;
          e2 = e2_i;
          score++;
        }
        b_tmp[a[i]] = score;
      }
      [b, b_tmp] = [b_tmp, b];
    }
    m *= 2;

    if (debug) {
      console.log(`m: ${m}, a:`, ...a, 'b:', ...b);
    }
  }

  return a;
};

const main = async (istr, ostr) => {
  const readline = require('readline');
  const it = readline.createInterface({input: istr})[Symbol.asyncIterator]();
  const getLine = async () => (await it.next()).value;
  const getMany = async () => (await getLine()).split(' ');

  const s = await getLine();
  const result = solve(s);
  ostr.write(`${result.join(' ')}\n`);
};

if (require.main === module) {
  main(process.stdin, process.stdout);
}

const test_0 = () => {
  const assert = require('assert');
  const s = 'ababba';
  assert.deepEqual(solve(s, 1), new Int32Array([6, 5, 0, 2, 4, 1, 3]));
};

const test_1 = () => {
  const inp = `
ababba
`.trimStart();
  const {Readable} = require('stream');
  main(Readable.from([inp]), process.stdout);
};

if (typeof module !== 'undefined') {
  module.exports = {test_0, test_1};
}
